src/react/Reconciler.js

```diff
@@ -0,0 +1,278 @@
+    realm: Realm,
+    moduleTracer: ModuleTracer,
+    statistics: ReactStatistics,
+    reactSerializerState: ReactSerializerState
+  ) {
+    this.realm = realm;
+    this.moduleTracer = moduleTracer;
+    this.statistics = statistics;
+    this.reactSerializerState = reactSerializerState;
+  }
+
+  realm: Realm;
+  moduleTracer: ModuleTracer;
+  statistics: ReactStatistics;
+  reactSerializerState: ReactSerializerState;
+
+  render(componentType: ECMAScriptSourceFunctionValue): Effects {
+    let propTypes = null;
+    let propsName = null;
+    let contextTypes = null;
+    let contextName = null;
+    // we take the first "props" paramater from "function MyComponent (props, context)" and look at its name
+    // if its not an Identifier, we leave propsName null so it doesn't get used to create the object
+    if (componentType.$FormalParameters.length > 0) {
+      if (t.isIdentifier(componentType.$FormalParameters[0])) {
+        propsName = ((componentType.$FormalParameters[0]: any): BabelNodeIdentifier).name;
+      }
+      // we expect that if there's a props argument, it should always have Flow annotations
+      invariant(
+        componentType.$FormalParameters[0].typeAnnotation,
+        `__registerReactComponentRoot() failed due to root component missing Flow type annotations for the "props" argument`
+      );
+      propTypes = flowAnnotationToObjectTypeTemplate(componentType.$FormalParameters[0].typeAnnotation);
+    }
+    if (componentType.$FormalParameters.length > 1) {
+      if (t.isIdentifier(componentType.$FormalParameters[1])) {
+        contextName = ((componentType.$FormalParameters[1]: any): BabelNodeIdentifier).name;
+      }
+      // we expect that if there's a context argument, it should always have Flow annotations
+      invariant(
+        componentType.$FormalParameters[1].typeAnnotation,
+        `__registerReactComponentRoot() failed due to root component missing Flow type annotations for the "context" argument`
+      );
+      contextTypes = flowAnnotationToObjectTypeTemplate(componentType.$FormalParameters[1].typeAnnotation);
+    }
+    return this.realm.wrapInGlobalEnv(() =>
+      this.realm.evaluateForEffects(() => {
+        let initialProps = createAbstractObject(this.realm, propsName, propTypes);
+        let initialContext = createAbstractObject(this.realm, contextName, contextTypes);
+        try {
+          let { result } = this._renderAsDeepAsPossible(
+            componentType,
+            initialProps,
+            initialContext,
+            BranchStatus.NO_BRANCH
+          );
+          this.statistics.optimizedTrees++;
+          return result;
+        } catch (e) {
+          // if there was a bail-out on the root component in this reconcilation process, then this
+          // should be an invariant as the user has explicitly asked for this component to get folded
+          invariant(false, "__registerReactComponentRoot() failed due to root component bailing out");
```

Comment:
If its a user error (Prepack is functioning correctly but the user input is unsupported) prefer throwing `CompilerDiagnostic` to an invariant failure. This way you can define a new error type and add a wiki page and we can link to that when the error is surfaced on the cli or repl. We use invariant failures to mean something internally has gone wrong in Prepack and there's a bug in Prepack that should be fixed.

See https://github.com/facebook/prepack/blob/master/src/serializer/modules.js#L147-L154 for an example.